### PR TITLE
Merge 2.12.x to 2.13.x [ci: last-only]

### PR DESCRIPTION
**Pull request by retronym**
Mostly skipped collections backports.

I skipped the benchmark in 94d8821ba1 as I beleive it overlaps with
`RedBlackTreeBenchmark` here, but maybe @mkeskells will find part of it
to forward port.

Some Tuple2 avoidance within Map operations from 03387fa2cb is skipped
but is adapted to 2.13.x and spun out to #8761.

```
S   *   348c541b8a (origin/2.12.x) Merge pull request #8736 from mkeskells/2.12.x_sortedSet-2
S   |\
S   | * e8e5fc8a2f (origin/pr/8736) a simple TreeSetBuilder
S   *   4516efdc5a Merge pull request #8742 from mkeskells/2.12.x_RedBlackBenckmark
S   |\
S   | * 393a0d5ca6 (origin/pr/8742) add more benchmarks for the RedBlackTrees TreeSet and TreeMap
S   | * 94d8821ba1 Basic TreeSet/TreeMap benchmark
S   *   90bc6cf275 Merge pull request #8737 from mkeskells/2.12.x_sortedSet-1
S   |\
S   | * 35c5194dba (origin/pr/8737) if its empty already then there is no need to create more garbage
S   | * 191340fe0e dont create a new TreeSet if there is no change to the underlying tree
S   *   b4c0225e8b Merge pull request #8738 from mkeskells/2.12.x_TreeMap-1
S   |\
S   | * 574758d518 (origin/pr/8738) Dont create a TreeMap if the underlying tree is unchanged
S   *   466f92f504 Merge pull request #8466 from rorygraves/mike/2.12.x_quickHashMap
S   |\
S   | * c0420576fd (rorygraves/mike/2.12.x_quickHashMap, retronym/review/8466, origin/pr/8466, review/8466) add HasForeachEntry
S   | * 4b91530a67 remove a couple of forwarding method
S   | * 03387fa2cb Improve performance of HashMap merge and ++ (mostly).
P   | * 85afb67931 add a benchmark
P   * cf929d678b Merge pull request #8688 from rorygraves/mike/2.12.x_toArray
P   * 6eb8576bb0 (rorygraves/mike/2.12.x_toArray, retronym/topic/review-toArray, origin/pr/8688, topic/review-toArray) Avoid generic array access in loop in ScalaRuntime
P   * e6befa227e Use a while loop in runtime support for varargs array conversions
P   * 2ef4752a7f (mike/2.12.x_toArray) don't create an IntRef in toArray
```

### CODE REVIEW

**Comment by lrytz** [comment]
```
sbt.ForkMain$ForkError: java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:267)
	at scala.collection.SetMapConsistencyTest.extraMutableAnyRefMapTests(SetMapConsistencyTest.scala:448)
```

(that test is also updated in https://github.com/scala/scala/pull/8761/files)

File: src/library/scala/collection/mutable/HashMap.scala [hunk 1]
```diff
@@ -102,7 +102,13 @@ class HashMap[K, V](initialCapacity: Int, loadFactor: Double)
           put0(next.key, next.value, next.hash, getOld = false)
         }
         this
-      case _ => super.addAll(xs)
+      case thatMap: Map[K, V] =>
+        thatMap.foreachEntry { (key: K, value: V) =>
+          put0(key, value, key.##, getOld = false)
```
Comment (szeiger) [inline src/library/scala/collection/mutable/HashMap.scala]: Here's the problem. You need `improveHash(key.##)`. The change comes from the commit titled "temp". It is not part of 2.12.x. Was this supposed to go into the separate PR for `foreachEntry`?

Comment (retronym) [reply src/library/scala/collection/mutable/HashMap.scala]: Yes, looks like I got my workspaces confused when also working on https://github.com/scala/scala/pull/8761/files#diff-5be35908b1ec9bbb90c66342b71dd9afR106

Comment (retronym) [reply src/library/scala/collection/mutable/HashMap.scala]: I've just added a new commit to revert `temp` because I'm never sure how to get git to DWIM when rebasing merges.

**Comment by szeiger** [comment]
Remaining 2.12 changes:

```
$ git log --oneline --graph 348c541b8a..upstream/2.12.x
*   e1e2bfeaa9 Merge pull request #8778 from retronym/ticket/sd674
|\
| * 98a5944d4d TreeMap is serialization compat with 2.12.10 again
* |   da1bf678e8 Merge pull request #8772 from hrhino/backport/8670
|\ \
| * | d4e90625fb [backport] Build intellij task: don't prompt for confirmation (#8670)
* | |   7ca43d5e7e Merge pull request #8776 from hrhino/topic/align
|\ \ \
| * | | c90c71634f Encourage alignment
| |/ /
* | |   ed743e5be5 Merge pull request #8770 from hrhino/t9908
|\ \ \
| * | | 552510050a Emit advice that compiles
|  / /
* | |   d1e9ac0cc2 Merge pull request #8769 from SethTisue/tweak-8732
|\ \ \
| |/ /
|/| |
| * | 48d3ee23e3 partially revert #8732 to remove problematic isEmpty checks
|/ /
* |   49cec62535 Merge pull request #8764 from retronym/topic/hashmap-merge-regression
|\ \
| |/
|/|
| * cf3a6f1874 Align {add,merge}{Common,Distinct}.
| * 0c62e21014 Fix regression in HashMap.{merge, ++}
|/
*   f139b61e8e Merge pull request #8759 from SethTisue/comment-mini-bootstrap
|\
| * 0469b597b8 improve doc of bootstrapping in PR validation on Travis-CI
*   154215bc18 Merge pull request #8755 from retronym/review/8735
|\
| * 398a891c02 Don't populate Context.implicits for Java sources
| * ca7dc4d597 Use Java semantics for imports in .java files
* ab4d97e87b Merge pull request #8732 from rorygraves/mike/2.12.x_quickTraversable
* 501533adeb remove some refs in TraversableOnce
```

501533adeb can be skipped, it no longer applies in 2.13.

@retronym I could use some help with the Context changes in 154215bc18. `importedSelectedSymbol` is completely different in 2.13.

**Comment by lrytz** [comment]
> @retronym I could use some help with the Context changes in [154215b](https://github.com/scala/scala/commit/154215bc18e6403d9d3b3777a7da53cad125e0cd). `importedSelectedSymbol` is completely different in 2.13.

If a change is hard to merge 2.12 -> 2.13, I think it's better to skip it (`-s ours`) and do / ask for a separate PR for 2.13. This makes it easier to review and avoids mistakes.

**Comment by retronym** [comment]
> ask for a separate PR for 2.13

https://github.com/scala/scala/pull/8784

**Comment by SethTisue** [comment]
is this good to go, then?

**Review by szeiger — APPROVED** [review]
LGTM. The skipped changes are not directly applicable to 2.13.x.
